src/ml/neural_net/combine_map.hpp

```diff
@@ -0,0 +1,147 @@
+/* Copyright © 2020 Apple Inc. All rights reserved.
+ *
+ * Use of this source code is governed by a BSD-3-clause license that can
+ * be found in the LICENSE.txt file or at
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+#ifndef ML_NEURAL_NET_COMBINE_MAP_HPP_
+#define ML_NEURAL_NET_COMBINE_MAP_HPP_
+
+#include <exception>
+#include <functional>
+#include <memory>
+#include <type_traits>
+
+#include <ml/neural_net/combine_base.hpp>
+
+namespace turi {
+namespace neural_net {
+
+template <typename T>
+class IteratorPublisher;
+
+/**
+ * Interface for objects that apply a transform to a value.
+ */
+template <typename T, typename U>
+class Transform {
+ public:
+  using Input = T;
+  using Output = U;
+
+  virtual ~Transform() = default;
+
+  /** Returns the next value in the sequence. May throw on error. */
+  virtual Output Invoke(Input value) = 0;
+};
+
+/**
+ * Templated implementation of Transform that wraps an arbitrary callable type.
+ */
+template <typename T, typename Callable>
+class CallableTransform
+    : public Transform<T, typename std::result_of<Callable(T)>::type> {
+ public:
+  using Input = T;
+  using Output = typename std::result_of<Callable(T)>::type;
+
+  CallableTransform(Callable impl) : impl_(std::move(impl)) {}
+
+  Output Invoke(Input input) override { return impl_(std::move(input)); }
+
+ private:
+  Callable impl_;
+};
+
+/**
+ * Concrete operator Publisher that wraps a Transform.
+ *
+ * The resulting Publisher inherits the semantics of the upstream Publisher that
+ * it subscribes to, with regard to the semantics of multiple downstream
```

Comment:
only one downstream.

Reply:
Just think twice. My bad, you can have multiple different subscribers subscribe to a publisher.